pkg/gatewayserver/io/ws/lbslns/upstream.go

```diff
@@ -303,17 +308,20 @@ func (req *JoinRequest) FromUplinkMessage(up *ttnpb.UplinkMessage, bandID string
 	return nil
 }
 
-// toUplinkMessage extracts fields from the LoRa Basics Station Uplink Data Frame "updf" message and converts them into an UplinkMessage for the network server.
-func (updf *UplinkDataFrame) toUplinkMessage(ids *ttnpb.GatewayIdentifiers, bandID string, receivedAt time.Time) (*ttnpb.UplinkMessage, error) {
-	var up ttnpb.UplinkMessage
+// toUplinkMessage extracts fields from the LoRa Basics Station Uplink Data Frame "updf"
+// message and converts them into an UplinkMessage for the network server.
+func (updf *UplinkDataFrame) toUplinkMessage(
+	ids *ttnpb.GatewayIdentifiers, bandID string, receivedAt time.Time,
+) (*ttnpb.UplinkMessage, error) {
+	up := &ttnpb.UplinkMessage{}
 	up.ReceivedAt = timestamppb.New(receivedAt)
 
 	parsedMHDR := &ttnpb.MHDR{}
 	if err := lorawan.UnmarshalMHDR([]byte{byte(updf.MHdr)}, parsedMHDR); err != nil {
 		return nil, errUplinkDataFrame.WithCause(err)
 	}
-	if parsedMHDR.MType != ttnpb.MType_UNCONFIRMED_UP && parsedMHDR.MType != ttnpb.MType_CONFIRMED_UP {
-		return nil, errMDHR.WithAttributes(`mhdr`, parsedMHDR)
+	if err := parsedMHDR.ValidateFields(); err != nil {
```

Comment:
This should be the place where we got an error before. `ValidateFields` will allow proprietary frames (as they are correct, protocol wise), but will reject zeroes or otherwise invalid values.